juju/main.py

```diff
@@ -0,0 +1,218 @@
+        endpoints = metadata.get(endpoint_type)
+        if endpoints:
+            charm_endpoints.extend(endpoints)
+    return charm_endpoints
+
+
+def _handle_event_link(charm_dir, event_name):
+    event_hook_path = os.path.join(
+        charm_dir,
+        'hooks', event_name.replace('_', '-'))
+
+    create_link = True
+    # Remove incorrect symlinks or files.
+    if os.path.exists(event_hook_path):
+        # Non-symlink entries and the ones not pointing to "install"
+        # need to be removed.
+        if not os.path.islink(event_hook_path):
+            debugf(f'Hook entry at {event_hook_path} is not a symlink:'
+                   ' attempting to remove it.')
+            # May raise OSError, e.g. in case it is a directory which
+            # is unexpected and left to the developer or operator to handle.
+            os.unlink(event_hook_path)
+        elif os.readlink(event_hook_path) != 'install':
+            debugf(f'Removing entry {event_hook_path} as it does not point'
+                   ' to "install"')
+            os.unlink(event_hook_path)
+        else:
+            create_link = False
+
+    if create_link:
+        debugf(f'Creating a new symlink named {event_hook_path}'
+               ' to hooks/install')
+        os.symlink('install', event_hook_path)
+
+
+def _setup_hooks(charm_dir, Charm, CharmEvents, Event, BoundEvent):
+    """Set up hooks for supported events.
+
+    Whether a charm can handle an event or not can be determined by
+    introspecting which events are bound to it.
+
+    charm_dir -- A root directory of the charm.
+    Charm -- A type of a concrete charm implementation.
+    CharmEvents -- A type declaring core Juju events.
+    Event -- A framework Event type.
+    BoundEvent -- A framework BoundEvent type.
+    """
+
+    relation_events = set()
+    for endpoint in _get_charm_endpoints():
+        relation_events.update(set([
+            f'{endpoint}_relation_{t}'
+            for t in ['joined', 'changed', 'departed', 'broken']
+        ]))
+
+    debugf(f'Discovered relation events from charm metadata:'
+           f' {relation_events}')
+
+    # Get core event names that CharmBase knows about.
+    core_events = set()
+    for event_name, instance in CharmEvents.__dict__.items():
+        if type(instance) == Event:
+            core_events.add(event_name)
+
+    # A set of all events that can possibly be supported based on core events
+    # and endpoints in the charm metadata.
+    possible_events = core_events.union(relation_events)
+
+    debugf(f'Discovered core charm events: {core_events}')
+
+    # Discover Juju events supported by the juju library
+    # using reflection.
+    supported_juju_events = set()
+    for event_name in dir(Charm.on):
+        attr_type = type(getattr(Charm.on, event_name))
+        if attr_type == BoundEvent and event_name in possible_events:
+            supported_juju_events.add(event_name)
```

Comment:
This can all be dropped in favor of `Charm.on.events()` now that my PR has landed.

Reply:
+1

Reply:
This makes me wonder if `EventsBase.events()` should in fact return a `dict` rather than a tuple generator.

Reply:
Will need this as well: https://github.com/canonical/charmbase/pull/9

Reply:
Updated #9 so this can become:

```python
supported_juju_events = {event_name for event_name in Charm.on.events()
                         if event_name in possible_events}
```